casper/src/test/scala/coop/rchain/casper/blocks/proposer/BlockReceiverStateSpec.scala

```diff
@@ -0,0 +1,104 @@
+
+    val (_, isReceiving) = newSt.beginStored("A1")
+    isReceiving shouldBe true
+  }
+
+  // endStored
+
+  "endStored" should "raise AssertionError if current state is not BeginStoreBlock" in {
+    val (st, _)    = BlockReceiverState[MId].beginStored("A1")
+    val (newSt, _) = st.endStored("A1", List.empty)
+
+    // A1 is now an EndStoreBlock but should be a BeginStoreBlock
+
+    assertThrows[AssertionError](newSt.endStored("A1", List.empty))
+  }
+
+  "endStored" should "update state of ending block, state of stored blocks and child relations" in {
+    val (st, _)                = BlockReceiverState[MId].beginStored("A2")
+    val (newSt, unseenParents) = st.endStored("A2", List("A1" -> true))
+
+    // Status of A2 changed from BeginStoreBlock to EndStoreBlock
+    st.receiveSt shouldNot contain("A2" -> EndStoreBlock)
+    newSt.receiveSt should contain("A2" -> EndStoreBlock)
+
+    // A1 is unseen parent so it is requested
+    st.receiveSt shouldNot contain("A1")
+    newSt.receiveSt should contain("A1" -> Requested)
+    unseenParents shouldBe Set("A1")
+
+    // Updated child-parent relations
+    newSt.blocksSt shouldBe Map("A2"       -> Set("A1"))
+    newSt.childRelations shouldBe Map("A1" -> Set("A2"))
+  }
+
+  // finished
+
+  "finished" should "raise AssertionError if block not in the state" in {
+    assertThrows[AssertionError](BlockReceiverState[MId].finished("A1", Set.empty))
+  }
+
+  "finished" should "raise AssertionError if block not received" in {
+    val (st, _) = BlockReceiverState[MId].beginStored("A1")
+    assertThrows[AssertionError](st.finished("A1", Set.empty))
+  }
+
+  "finished" should "remove resolved deps and return set of blocks with resolved deps" in {
+    // Started storing of A2
+    val (st1, _) = BlockReceiverState[MId].beginStored("A2")
+    st1.receiveSt shouldBe Map("A2" -> BeginStoreBlock)
+
+    // Finished storing of A2 with unseen parent A1
+    val (st2, a2UnseenParents) = st1.endStored("A2", List("A1" -> true))
+    st2.blocksSt shouldBe Map("A2" -> Set("A1"))
+    st2.receiveSt should contain allOf ("A2" -> EndStoreBlock, "A1" -> Requested)
+    st2.childRelations shouldBe Map("A1" -> Set("A2"))
+    a2UnseenParents shouldBe Set("A1")
+
+    // Started storing of A1
+    val (st3, _) = st2.beginStored("A1")
+    st3.receiveSt should contain("A1" -> BeginStoreBlock)
+
+    // Finished storing of A1 without parents
+    val (st4, a1UnseenParents) = st3.endStored("A1", List.empty)
+    st4.blocksSt should contain("A1"  -> Set())
+    st4.receiveSt should contain("A1" -> EndStoreBlock)
+    a1UnseenParents shouldBe empty
+
+    // After the finishing of A1, it will be removed from receive state
+    // Child A2 has dependencies resolved and is pending validation
+    val (st5, depsValidated) = st4.finished("A1", Set.empty)
+    st5.receiveSt shouldNot contain("A1")
+    st5.receiveSt should contain("A2" -> PendingValidation)
+    depsValidated shouldBe Set("A2")
+  }
```

Comment:
One property which receiver should satisfy is to get to the empty state if all blocks are processed/validated. If something lefts in the state we can consider as memory leak because logic requires to remove all trace of blocks after they are validated.
Please add additional tests to prove this property.

Reply:
[Done](https://github.com/rchain/rchain/pull/3738/commits/2bff27b21e314186bdfd09d153b9e873a70b44cb).